File: DataStructures/Linear/LinkedList/ex.py
class Node:
  def __init__(self, value):
    self.value = value
    self.next = None

class LinkedList:
  def __init__(self):
    self.process = True
    self.head = None

  def visualize(self):
    link = self.head

    while link:
      print(link.value, end='')
      link = link.next
      print(' > ', end='')

    print('NULL\n')

  def length(self):
    link, count = self.head, 0
    while link: 
      link = link.next
      count += 1
    return count

  def add(self, value):
    if self.process: 
      print(f'add({value})')

    node = Node(value)

    if self.head:
      cur = self.head
      temp = None

      while cur:
        temp = cur
        cur = cur.next

      temp.next = node

    else:
      self.head = node

    if self.process: 
      self.visualize()

  def insert(self, index, value):
    if not (0 <= index < self.length()):
      print('[ERROR] Insert failed.')
      print(f'Index out of range: 0 ~ {self.length()} but "{index}"\n')
      return

    if self.process: 
      print(f'insert({index}, {value})')

    node = Node(value)
    cur = self.head

    if index == 0:
      node.next = cur
      self.head = node

    elif index == self.length() - 1:
      self.add(value)
      return

    for count in range(1, self.length()):
      if index == count:
        node.next = cur.next
        cur.next = node
        break

      cur = cur.next

    if self.process:
      self.visualize()

  def pop(self):
    if not self.head:
      print('[ERROR] Pop failed.')
      print('List is Empty')
      return
    
    print('pop()', end=' = ')

    rm_node = None
    cur = self.head

    while cur:
      if not cur.next:
        rm_node = self.head
        self.head = None
        break

      if not cur.next.next:
        rm_node = cur.next
        cur.next = None
        break

      cur = cur.next
    
    print(rm_node.value)

    if self.process: 
      self.visualize()

  def remove(self, index):
    if not self.head:
      print('[ERROR] Remove failed.')
      print('List is Empty')
      return
    
    if not (0 <= index < self.length()):
      print('[ERROR] Remove failed.')
      print(f'Index out of range: 0 ~ {self.length()} but "{index}"\n')
      return

    print(f'remove({index})', end=' = ')

    rm_node = None
    cur = self.head

    if index == 0:
      rm_node = cur
      self.head = cur.next

    elif index == self.length() - 1:
      self.pop()
      return

    for count in range(1, self.length()):
      if index == count:
        rm_node = cur.next
        cur.next = cur.next.next
        break

      cur = cur.next

    print(rm_node.value)

    if self.process: 
      self.visualize()

    return rm_node.value
  
def choose():
  print()
  print('== Menu ==')
  print('[0] Quit')
  print('[1] Add')
  print('[2] Insert')
  print('[3] Pop')
  print('[4] Remove')
  print('[5] Visualize')
  print('>>', end=' ')
  return int(input())

if __name__ == '__main__':
  l = LinkedList()

  while ipt := choose():
    print()

    try:
      func, param = [
        ['add', 1], 
        ['insert', 2], 
        ['pop', 0], 
        ['remove', 1],
        ['visualize', 0],
      ][ipt - 1]
      print(f'{func}(param={param})')
      
      params = []
      if param: 
        print('>>', end=' ')
        params = input().split()
        print()

        if len(params) != param:
          print('[ERROR] Invalid input')
          continue
      
      eval(f'l.{func}({",".join(params)})')

    except:
      print('[ERROR] Invalid input')
      continue

File: DataStructures/Linear/LinkedList/ex.cpp
#include <iostream>
#include <string>

using namespace std;

struct node {
    int data;
    node* next_node;
};

class LinkedList {
private:
    node* head;
    node* tail;
public:
    LinkedList() {
        head = NULL;
        tail = NULL;
    }
    node* getHead();
    void add(int n);
    bool insert(int index, int n);
    bool pop();
    bool remove(int index);
    void display(node* head);
};

node* LinkedList::getHead() {
    return head;
}

void LinkedList::add(int n) {
    node* new_node = new node;
    new_node->data = n;

    if(head == NULL) {
        head = new_node;
        tail = new_node;
    } else {
        tail->next_node = new_node;
        tail = new_node;
    }
}

bool LinkedList::insert(int index, int n) {
    node* new_node = new node;
    new_node->data = n;

    if(index == 0) {
        new_node->next_node = head;
        head = new_node;
        if(tail == NULL) {
            tail = new_node;
        } 
     } else {
        node* cur_node = head;
        for(int i = 0; i < index - 1; i++) {
            if(cur_node == NULL) {
                cout << "\nTry another index (out of range)\n";
                delete new_node;
                return false;
            }
            cur_node = cur_node->next_node;
        }
        new_node->next_node = cur_node->next_node;
        cur_node->next_node = new_node;
        if(new_node->next_node == NULL) 
            tail = new_node;
    }
    return true;
}

bool LinkedList::pop() {
    if(head == NULL) {
        cout << "\nLinkedList is empty\n";
        return false;
    } else if(head == tail) {
        delete tail;
        head = tail = NULL;
    } else {
        node* prv_node = head;
        while(prv_node->next_node != tail) {
            prv_node = prv_node->next_node;
        }
        delete tail;
        tail = prv_node;
        tail->next_node = NULL;
    }
    return true;
}

bool LinkedList::remove(int index) {
    if(head == NULL) {
        cout << "LinkedList is empty\n";
        return false;
     } else if(index == 0) {
        node* temp_node = head->next_node;
        delete head;
        head = temp_node;

        if(head == NULL)
            tail = NULL;
     } else {
        node* cur_node = head;
        for(int i = 0; i < index - 1; i++) {
            if(cur_node == NULL) {
                cout << "\nTry another index (out of range)\n";
                return false;
            }
            cur_node = cur_node->next_node;
        }
        node* temp_node = cur_node->next_node;
        cur_node->next_node = temp_node->next_node;
        delete temp_node;

        if(temp_node->next_node == NULL) 
            tail = temp_node;
    }
    return true;
}

void LinkedList::display(node* head) {
    
    if(head == NULL)
        cout << "]\n";
    else {
        cout << head->data << ", ";
        display(head->next_node);
    }
}

int main() {
    LinkedList lList;

    while(true) {
        bool check = true;

        cout << "\n=== Linked List Tester ===\n" 
        << "(1) Add node\n"
        << "(2) Insert node\n"
        << "(3) Pop node\n"
        << "(4) Remove node\n"
        << "(5) Display node\n"
        << "(0) Exit\n"
        << "==========================\n";

        int selectNum = 0;
        cout << "\nSelect your action: ";
        cin >> selectNum;
        cout << "\n";

        if(selectNum == 1) {
            int num = 0;
            cout << "Add num? -> ";
            cin >> num;
            
            lList.add(num);
            cout << "\nAdd Completed!\n";
        } else if(selectNum == 2) {
            int index = 0, num = 0;
            cout << "\nIndex? -> ";
            cin >> index;
            cout << "Add num? -> ";
            cin >> num;

            check = lList.insert(index, num);
            if(check)
                cout << "\nInsert Completed!\n";
            else
                cout << "\nInsert Failed!\n";
        } else if(selectNum == 3) {
            check = lList.pop();
            if(check)
                cout << "\nPop Completed!\n";
            else
                cout << "\nPop Failed!\n";
        } else if(selectNum == 4) {
            int index = 0;
            cout << "\nIndex? -> ";
            cin >> index;

            check = lList.remove(index);
            if(check)
                cout << "\nRemove Completed!\n";
            else
                cout << "\nRemove Failed!\n";
        } else if(selectNum == 5) {
            cout << "++ My Linked List Data ++\n" << "[ ";
            
            lList.display(lList.getHead());
        } else {
            cout << "Tester Shut down! Bye ~\n\n";
            break;
        }
    }
}